### PR TITLE
Use `assert_same` and `refute_same` instead of `equal?`

### PR DESCRIPTION
**Pull request by tejasbubane**
https://github.com/rubocop/rubocop-minitest/issues/215

### CODE REVIEW

**Comment by andyw8** [comment]
Repeating my reply from https://github.com/rubocop/rubocop-minitest/issues/215 for visibility:

I have a slight concern here: `same` is less explicit than `equal?`. It's not obvious which kind of comparison is being made (`==`, `eql?`, etc. ) unless you look at the implementation or docs.

I think it could be mitigated with something like:

`...unless there is a

File: README.adoc [hunk 1]
```diff
@@ -118,6 +118,33 @@ assert(!"rubocop-minitest" == (actual))
 refute_equal("rubocop-minitest", actual)
 ----
 
+=== Assert Same [[assert-same]]
+
+Use `assert_same` instead of `assert` + `equal?`.
```
Comment (andyw8) [inline README.adoc:123]:
```suggestion
Use `assert_same` instead of `assert` with `equal?`.
```

File: README.adoc [hunk 2]
```diff
@@ -118,6 +118,33 @@ assert(!"rubocop-minitest" == (actual))
...
+# bad
+assert foo.equal?(bar)
+
+# good
+assert_same foo, bar
+----
+
+=== Refute Same [[refute-same]]
+
+Use `refute_same` instead of `refute` + `equal?`.
```
Comment (andyw8) [inline README.adoc:136]:
```suggestion
Use `refute_same` instead of `refute` with `equal?`.
```

**Comment by tejasbubane** [comment]
> I have a slight concern here: same is less explicit than equal?. It's not obvious which kind of comparison is being made (==, eql?, etc. ) unless you look at the implementation or docs.

This applies to `assert_equal` too. To begin with `==`, `eql?`, `equal?`, etc are equally confusing for newcomers.  I feel `same` means same object which is checked by `equal?`.

**Comment by andyw8** [comment]
That's true. But `assert_equal` is much more common, so it's better understood.

I'm not against this as a new cop, but there's some nuance, and I'm not sure how best to handle it.

**Comment by koic** [comment]
@andyw8's concerns are understandable. However, `equal?` is for confirming object identity, and "`asssert_same`" seems to be a reasonable naming. After all, it is a point to keep in mind the difference between object identity comparison and object value comparison. I remember having a similar talk when writing the following rule:
https://rubystyle.guide/#identity-comparison

File: README.adoc [hunk 3]
```diff
@@ -118,6 +118,33 @@ assert(!"rubocop-minitest" == (actual))
...
+
+Use `assert_same` instead of `assert` with `equal?`.
+
+[source,ruby]
+----
+# bad
+assert foo.equal?(bar)
+
+# good
+assert_same foo, bar
```
Comment (koic) [inline README.adoc:131]:
```suggestion
# bad
assert(expected.equal?(actual))

# good
assert_same(expected, actual)
```

File: README.adoc [hunk 4]
```diff
@@ -118,6 +118,33 @@ assert(!"rubocop-minitest" == (actual))
...
+Use `refute_same` instead of `refute` with `equal?`.
+
+[source,ruby]
+----
+# bad
+refute foo.equal?(bar)
+assert !foo.equal?(bar)
+
+# good
+refute_same foo, bar
```
Comment (koic) [inline README.adoc:145]:
```suggestion
# bad
refute(expected.equal?(actual))
assert(!expected.equal?(actual))

# good
refute_same(expected, actual)
```

File: README.adoc [hunk 5]
```diff
@@ -118,6 +118,33 @@ assert(!"rubocop-minitest" == (actual))
 refute_equal("rubocop-minitest", actual)
 ----
 
+=== Assert Same [[assert-same]]
+
+Use `assert_same` instead of `assert` with `equal?`.
+
```
Comment (koic) [inline README.adoc:124]: It may be convenient for users to write a `NOTE` about the difference in comparison method with `assert_equal`.

Comment (andyw8) [reply README.adoc:124]: Suggestion:

> Use `assert_same` only when there is a need to compare by identity. Otherwise, use `assert_equal`.

**Review by andyw8 — APPROVED** [review]

**Pull request merged**